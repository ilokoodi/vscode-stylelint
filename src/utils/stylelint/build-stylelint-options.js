'use strict';

const path = require('path');
const pathIsInside = require('path-is-inside');
const { URI } = require('vscode-uri');
const { findPackageRoot } = require('../packages');

/**
 * Given a document URI, base options, and extension options, builds a Stylelint
 * options object. Extension options supersede base options.
 * @param {string} uri
 * @param {string} [workspaceFolder]
 * @param {Partial<stylelint.LinterOptions>} [baseOptions]
 * @param {ExtensionOptions} [extensionOptions]
 * @returns {Promise<Partial<stylelint.LinterOptions>>}
 */
async function buildStylelintOptions(
	uri,
	workspaceFolder,
	baseOptions = {},
	{
		config,
		configFile,
		configBasedir,
		customSyntax,
		ignoreDisables,
		reportNeedlessDisables,
		reportInvalidScopeDisables,
	} = {},
) {
	const options = {
		...baseOptions,

		config: config ?? baseOptions.config,

		configFile: configFile
			? workspaceFolder
				? configFile.replace(/\$\{workspaceFolder\}/gu, workspaceFolder)
				: configFile
			: baseOptions.configFile,

		configBasedir: configBasedir
			? path.isAbsolute(configBasedir)
				? configBasedir
				: path.join(workspaceFolder ?? '', configBasedir)
			: baseOptions.configBasedir,

		customSyntax: customSyntax
			? workspaceFolder
				? customSyntax.replace(/\$\{workspaceFolder\}/gu, workspaceFolder)
				: customSyntax
			: baseOptions.customSyntax,

		ignoreDisables: ignoreDisables ?? baseOptions.ignoreDisables,

		reportNeedlessDisables: reportNeedlessDisables ?? baseOptions.reportNeedlessDisables,

		reportInvalidScopeDisables:
			reportInvalidScopeDisables ?? baseOptions.reportInvalidScopeDisables,
	};

	const documentPath = URI.parse(uri).fsPath;

	if (documentPath) {
		if (workspaceFolder && pathIsInside(documentPath, workspaceFolder)) {
			options.ignorePath = path.join(workspaceFolder, '.stylelintignore');
		}

		if (options.ignorePath === undefined) {
			options.ignorePath = path.join(
				(await findPackageRoot(documentPath)) || path.parse(documentPath).root,
				'.stylelintignore',
			);
		}
	}

	return options;
}

module.exports = {
	buildStylelintOptions,
};
