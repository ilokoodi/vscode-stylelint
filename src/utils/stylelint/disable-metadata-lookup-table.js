'use strict';

const { intersect } = require('../sets');
const { getDisableDiagnosticRule } = require('./get-disable-diagnostic-rule');

/**
 * Helps lookup disable reports by type, rule, or range.
 */
class DisableMetadataLookupTable {
	/**
	 * Reports by type.
	 * @type {Map<string, Set<lsp.Diagnostic>>}
	 */
	#reportsByType = new Map();

	/**
	 * Reports by type.
	 * @type {Map<string, Set<lsp.Diagnostic>>}
	 */
	#reportsByRule = new Map();

	/**
	 * Reports by type.
	 * @type {Map<string, Set<lsp.Diagnostic>>}
	 */
	#reportsByRange = new Map();

	/**
	 * @param {lsp.Diagnostic[]} diagnostics The diagnostics to build the lookup
	 * table from.
	 */
	constructor(diagnostics) {
		for (const diagnostic of diagnostics) {
			const rule = getDisableDiagnosticRule(diagnostic);

			if (!rule) {
				continue;
			}

			// If getDisableDiagnosticRule returns a rule, the diagnostic code
			// must be a string.
			const code = /** @type {string} */ (diagnostic.code);

			const existingByType = this.#reportsByType.get(code);

			if (existingByType) {
				existingByType.add(diagnostic);
			} else {
				this.#reportsByType.set(code, new Set([diagnostic]));
			}

			const existingByRule = this.#reportsByRule.get(rule);

			if (existingByRule) {
				existingByRule.add(diagnostic);
			} else {
				this.#reportsByRule.set(rule, new Set([diagnostic]));
			}

			const rangeKey = this.#getRangeKey(diagnostic.range);

			const existingByRange = this.#reportsByRange.get(rangeKey);

			if (existingByRange) {
				existingByRange.add(diagnostic);
			} else {
				this.#reportsByRange.set(rangeKey, new Set([diagnostic]));
			}
		}
	}

	/**
	 * @param {lsp.Range} range
	 * @returns {string}
	 */
	#getRangeKey({ start, end }) {
		return `${start.line}:${start.character}:${end.line}:${end.character}`;
	}

	/**
	 * Finds the reports that match the given type, rule, and range.
	 * @param {{
	 *   type?: import('../types').DisableReportRuleNames | string;
	 *   rule?: string;
	 *   range?: lsp.Range;
	 * }} options
	 * @returns {Set<lsp.Diagnostic>}
	 */
	find({ type, rule, range }) {
		const reportsByType = type ? this.#reportsByType.get(type) : undefined;
		const reportsByRule = rule ? this.#reportsByRule.get(rule) : undefined;
		const reportsByRange = range ? this.#reportsByRange.get(this.#getRangeKey(range)) : undefined;

		return intersect(intersect(reportsByType, reportsByRule), reportsByRange) ?? new Set();
	}
}

module.exports = {
	DisableMetadataLookupTable,
};
