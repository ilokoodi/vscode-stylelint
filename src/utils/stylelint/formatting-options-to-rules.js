'use strict';

/**
 * @typedef {{
 *   indentation: [number | string],
 *   'no-missing-end-of-source-newline'?: boolean,
 *   'no-eol-whitespace'?: boolean,
 * }} FormattingRules
 */

/**
 * Converts the given formatting options to rules.
 * @param {lsp.FormattingOptions} options The formatting options.
 * @returns {FormattingRules} The rules.
 */
function formattingOptionsToRules({
	insertSpaces,
	tabSize,
	insertFinalNewline,
	trimTrailingWhitespace,
}) {
	// NOTE: There is no equivalent rule for trimFinalNewlines, so it is not respected.
	// TODO: Create respective rule upstream?

	/** @type {FormattingRules} */
	const rules = {
		indentation: [insertSpaces ? tabSize : 'tab'],
	};

	if (insertFinalNewline !== undefined) {
		rules['no-missing-end-of-source-newline'] = insertFinalNewline;
	}

	if (trimTrailingWhitespace !== undefined) {
		rules['no-eol-whitespace'] = trimTrailingWhitespace;
	}

	return rules;
}

module.exports = {
	formattingOptionsToRules,
};
