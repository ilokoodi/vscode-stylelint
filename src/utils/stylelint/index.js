'use strict';

module.exports = {
	...require('./build-stylelint-options'),
	...require('./disable-metadata-lookup-table'),
	...require('./formatting-options-to-rules'),
	...require('./get-disable-diagnostic-rule'),
	...require('./process-linter-result'),
	...require('./stylelint-runner'),
	...require('./warning-to-diagnostic'),
};
