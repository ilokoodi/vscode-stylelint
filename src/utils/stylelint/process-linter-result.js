'use strict';

const { warningToDiagnostic } = require('./warning-to-diagnostic');
const { InvalidOptionError } = require('../types');

/**
 * Processes the results of a Stylelint lint run.
 *
 * If Stylelint reported any warnings, they are converted to Diagnostics and
 * returned. If the lint results contain raw output in the `output` property, it
 * is also returned.
 *
 * Throws an `InvalidOptionError` for any invalid option warnings reported by
 * Stylelint.
 * @param {stylelint.PublicApi} stylelint The Stylelint instance that was used.
 * @param {stylelint.LinterResult} result The results returned by Stylelint.
 * @returns {LintDiagnostics}
 */
function processLinterResult(stylelint, { results, output }) {
	if (results.length === 0) {
		return { diagnostics: [] };
	}

	const [{ invalidOptionWarnings, warnings, ignored }] = results;

	if (ignored) {
		return { diagnostics: [] };
	}

	if (invalidOptionWarnings.length !== 0) {
		throw new InvalidOptionError(invalidOptionWarnings);
	}

	const diagnostics = warnings.map((warning) => warningToDiagnostic(warning, stylelint.rules));

	return output ? { output, diagnostics } : { diagnostics };
}

module.exports = {
	processLinterResult,
};
