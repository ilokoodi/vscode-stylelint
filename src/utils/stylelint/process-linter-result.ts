import { warningToDiagnostic } from './warning-to-diagnostic';
// eslint-disable-next-line n/no-unpublished-import
import type { LinterResult } from 'stylelint';
import { type LintDiagnostics, type Stylelint, InvalidOptionError } from './types';

/**
 * Processes the results of a Stylelint lint run.
 *
 * If Stylelint reported any warnings, they are converted to Diagnostics and
 * returned. If the lint results contain raw output in the `output` property, it
 * is also returned.
 *
 * Throws an `InvalidOptionError` for any invalid option warnings reported by
 * Stylelint.
 * @param stylelint The Stylelint instance that was used.
 * @param result The results returned by Stylelint.
 */
export function processLinterResult(
	stylelint: Stylelint,
	linterResult: LinterResult,
): LintDiagnostics {
	const { results } = linterResult;

	if (results.length === 0) {
		return { diagnostics: [] };
	}

	const [{ invalidOptionWarnings, warnings, ignored }] = results;

	if (ignored) {
		return { diagnostics: [] };
	}

	if (invalidOptionWarnings.length !== 0) {
		throw new InvalidOptionError(invalidOptionWarnings);
	}

	let { ruleMetadata } = linterResult;

	if (!ruleMetadata) {
		// Create built-in rule metadata for backwards compatibility.
		ruleMetadata = new Proxy(
			{},
			{
				get: (_, key: string) => {
					// eslint-disable-next-line @typescript-eslint/ban-ts-comment
					// @ts-ignore -- (TS7053) `stylelint.rules` has returned `Promise` values since v16.
					// See https://stylelint.io/migration-guide/to-16#changed-nodejs-api-stylelintrules-object
					//
					// eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-member-access
					return stylelint.rules?.[key]?.meta;
				},
			},
		);
	}

	const diagnostics = warnings.map((warning) => warningToDiagnostic(warning, ruleMetadata));
	const output = ('report' in linterResult && linterResult.report) || linterResult.output;

	return output ? { output, diagnostics } : { diagnostics };
}
