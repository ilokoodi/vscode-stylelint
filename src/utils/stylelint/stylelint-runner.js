'use strict';

const os = require('os');
const { URI } = require('vscode-uri');
const { StylelintResolver } = require('../packages');
const { getWorkspaceFolder } = require('../documents');
const { processLinterResult } = require('./process-linter-result');
const { buildStylelintOptions } = require('./build-stylelint-options');

/**
 * Runs Stylelint in VS Code.
 */
class StylelintRunner {
	/**
	 * The language server connection.
	 * @type {lsp.Connection | undefined}
	 */
	#connection;

	/**
	 * The logger to use, if any.
	 * @type {winston.Logger | undefined}
	 */
	#logger;

	/**
	 * The Stylelint resolver.
	 * @type {StylelintResolver}
	 */
	#stylelintResolver;

	/**
	 * @param {lsp.Connection} [connection]
	 * @param {winston.Logger} [logger]
	 * @param {StylelintResolver} [resolver]
	 */
	constructor(connection, logger, resolver) {
		this.#connection = connection;
		this.#logger = logger;
		this.#stylelintResolver = resolver ?? new StylelintResolver(connection, logger);
	}

	/**
	 * Lints the given document using Stylelint. The linting result is then
	 * converted to LSP diagnostics and returned.
	 * @param {lsp.TextDocument} document
	 * @param {stylelint.LinterOptions} [linterOptions]
	 * @param {ExtensionOptions} [extensionOptions]
	 * @returns {Promise<LintDiagnostics>}
	 */
	async lintDocument(document, linterOptions = {}, extensionOptions = {}) {
		const workspaceFolder =
			this.#connection && (await getWorkspaceFolder(this.#connection, document));

		const result = await this.#stylelintResolver.resolve(extensionOptions, document);

		if (!result) {
			this.#logger?.info('No Stylelint found with which to lint document', {
				uri: document.uri,
				options: extensionOptions,
			});

			return { diagnostics: [] };
		}

		const { stylelint } = result;

		const { fsPath } = URI.parse(document.uri);

		// Workaround for Stylelint treating paths as case-sensitive on Windows
		// If the drive letter is lowercase, we need to convert it to uppercase
		// See https://github.com/stylelint/stylelint/issues/5594
		// TODO: Remove once fixed upstream
		const codeFilename =
			os.platform() === 'win32'
				? fsPath.replace(/^[a-z]:/, (match) => match.toUpperCase())
				: fsPath;

		/** @type {stylelint.LinterOptions} */
		const options = {
			...(await buildStylelintOptions(
				document.uri,
				workspaceFolder,
				linterOptions,
				extensionOptions,
			)),
			code: document.getText(),
			formatter: () => '',
		};

		if (codeFilename) {
			options.codeFilename = codeFilename;
		} else if (!linterOptions?.config?.rules) {
			options.config = { rules: {} };
		}

		if (this.#logger?.isDebugEnabled()) {
			this.#logger?.debug('Running Stylelint', { options: { ...options, code: '...' } });
		}

		try {
			return processLinterResult(stylelint, await stylelint.lint(options));
		} catch (err) {
			if (
				err instanceof Error &&
				(err.message.startsWith('No configuration provided for') ||
					err.message.includes('No rules found within configuration'))
			) {
				// Check only CSS syntax errors without applying any Stylelint rules
				return processLinterResult(
					stylelint,
					await stylelint.lint({ ...options, config: { rules: {} } }),
				);
			}

			throw err;
		}
	}
}

module.exports = {
	StylelintRunner,
};
