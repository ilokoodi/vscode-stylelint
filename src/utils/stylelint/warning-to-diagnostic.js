'use strict';

const { Diagnostic, DiagnosticSeverity, Position, Range } = require('vscode-languageserver-types');

/**
 * Converts a Stylelint warning to an LSP Diagnostic.
 *
 * @example
 * ```js
 * const [result] = await stylelint.lint({
 *   code: 'a { color: red; }',
 *   config: { rules: { 'color-named': 'never' } }
 * });
 *
 * const [warning] = result.warnings;
 * // {
 * //   rule: 'color-named',
 * //   text: 'Unexpected named color "red" (color-named)',
 * //   severity: 'error',
 * //   line: 1,
 * //   column: 12
 * // }
 *
 * const diagnostic = warningToDiagnostic(warning);
 * // {
 * //   message: 'Unexpected named color "red" (color-named)',
 * //   severity: 1,
 * //   source: 'Stylelint',
 * //   range: {
 * //     start: {
 * //       line: 0,
 * //       character: 11
 * //     },
 * //     end: {
 * //       line: 0,
 * //       character: 11
 * //     }
 * //   }
 * // }
 * ```
 * @param {stylelint.Warning} warning The warning to convert.
 * @param {{[name: string]: stylelint.Rule}} [rules] Available Stylelint rules.
 * @returns {Diagnostic}
 */
function warningToDiagnostic(warning, rules) {
	const position = Position.create(warning.line - 1, warning.column - 1);

	const ruleDocUrl =
		rules?.[warning.rule] && `https://stylelint.io/user-guide/rules/${warning.rule}`;

	const diagnostic = Diagnostic.create(
		Range.create(position, position),
		warning.text,
		DiagnosticSeverity[warning.severity === 'warning' ? 'Warning' : 'Error'],
		warning.rule,
		'Stylelint',
	);

	if (ruleDocUrl) {
		diagnostic.codeDescription = { href: ruleDocUrl };
	}

	return diagnostic;
}

module.exports = {
	warningToDiagnostic,
};
