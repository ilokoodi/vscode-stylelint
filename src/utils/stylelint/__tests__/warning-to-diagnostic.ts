import { lint } from 'stylelint';
import { warningToDiagnostic } from '../warning-to-diagnostic';
import type stylelint from 'stylelint';

describe('warningToDiagnostic', () => {
	test('should convert a Stylelint warning to an LSP diagnostic', async () => {
		expect.assertions(2);
		const {
			results: [{ warnings }],
		} = await lint({
			code: `a {
				color: #AAA;
				border-color: #bbbbbb;
			}`,
			config: {
				rules: {
					'color-hex-case': ['lower'],
					'color-hex-length': ['short', { severity: 'warning' }],
				},
			},
		});

		expect(warningToDiagnostic(warnings[0])).toMatchSnapshot();
		expect(warningToDiagnostic(warnings[1])).toMatchSnapshot();
	});

	test('should add a rule documentation URL if a matching rule exists', async () => {
		expect.assertions(1);
		const {
			results: [{ warnings }],
		} = await lint({
			code: `a {
				color: #AAA;
				border-color: #bbbbbb;
			}`,
			config: {
				rules: {
					'color-hex-case': ['lower'],
				},
			},
		});

		const rules = {
			'color-hex-case': { meta: { url: 'https://stylelint.io/rules/color-hex-case' } },
		} as unknown as {
			[name: string]: stylelint.Rule<unknown, unknown>;
		};

		expect(warningToDiagnostic(warnings[0], rules)).toMatchSnapshot();
	});
});
