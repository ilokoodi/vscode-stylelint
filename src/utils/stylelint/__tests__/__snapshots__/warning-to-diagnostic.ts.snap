// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`warningToDiagnostic should add a rule documentation URL if a matching rule exists 1`] = `
Object {
  "code": "color-no-invalid-hex",
  "codeDescription": Object {
    "href": "https://stylelint.io/rules/color-no-invalid-hex",
  },
  "message": "Unexpected invalid hex color \\"#AA\\" (color-no-invalid-hex)",
  "range": Object {
    "end": Object {
      "character": 14,
      "line": 1,
    },
    "start": Object {
      "character": 11,
      "line": 1,
    },
  },
  "severity": 1,
  "source": "Stylelint",
}
`;

exports[`warningToDiagnostic should convert a Stylelint warning to an LSP diagnostic 1`] = `
Object {
  "code": "color-hex-length",
  "message": "Expected \\"#bbbbbb\\" to be \\"#bbb\\" (color-hex-length)",
  "range": Object {
    "end": Object {
      "character": 25,
      "line": 2,
    },
    "start": Object {
      "character": 18,
      "line": 2,
    },
  },
  "severity": 2,
  "source": "Stylelint",
}
`;

exports[`warningToDiagnostic should convert a Stylelint warning to an LSP diagnostic 2`] = `
Object {
  "code": "color-no-invalid-hex",
  "message": "Unexpected invalid hex color \\"#AA\\" (color-no-invalid-hex)",
  "range": Object {
    "end": Object {
      "character": 14,
      "line": 1,
    },
    "start": Object {
      "character": 11,
      "line": 1,
    },
  },
  "severity": 1,
  "source": "Stylelint",
}
`;
